### PR TITLE
NF: rumba reconst

### PR DESCRIPTION
**Pull request by kenjimarshall**
Robust and Unbiased Model-BAsed Spherical Deconvolution (RUMBA-SD)

- RumbaSD and RumbaFit classes for local reconstruction
- Standalone global_fit method for whole brain fitting with built-in spatial regularization option
- Examples for both local reconstruction (single- and multi-shell) and fiber tracking with RUMBA-SD
- Model compartments for white matter, grey matter, and cerebrospinal fluid

Developed in collaboration with @gabknight and @ejcanalesr. RUMBA-SD is the algorithm that won the MICCAI IronTract 2019 challenge, the 3D VoTEM ISBI 2018 challenge (sub-challenge #2), and the co-winner of the ISBI HARDI 2013 challenge.

Reference:

Canales-Rodríguez, E. J., Daducci, A., Sotiropoulos, S. N., Caruyer, E., Aja-Fernández, S., Radua, J., Mendizabal, J. M. Y., Iturria-Medina, Y., Melie-García, L., Alemán-Gómez, Y., Thiran, J.-P., Sarró, S., Pomarol-Clotet, E., & Salvador, R. (2015). Spherical Deconvolution of Multichannel Diffusion MRI Data with Non-Gaussian Noise Models and Spatial Regularization. PLOS ONE, 10(10), e0138910. https://doi.org/10.1371/journal.pone.0138910

### CODE REVIEW

**Comment by Garyfallidis** [comment]
Exciting to see this here @kenjimarshall, @ejcanalesr and @gabknight ! Please also provide some insights from comparisons with other methods so we can understand impact. We will need a bit more specific information than what is provided above. So why is this method more advantageous than existing techniques and how much more advantageous? Any disadvantages should be reported too.

**Comment by Garyfallidis** [comment]
Please also add the reference of the paper in the description.

**Comment by gabknight** [comment]
Hi @Garyfallidis, RUMBA is a dictionary-based deconvolution algorithm. It is thus comparable to csdeconv and mcsd. It works best using multi-shell diffusion data, but can be parametrised for single shell data too. Once fitted, we have access to the ODF and to the fraction of isotropic diffusion water. It is implemented following the dipy reconst model/fit framework. It works in the discrete signal/sphere space (not in SH).

Additionally, @kenjimarshall has implemented the RUMBA global fit (`dipy.reconst.rumba.global_fit(.)`), which fits the whole volume at once using total variation spatial regularisation. This is the fit method that has been successful in reconstruction challenges over the past few years. The code is based on the Matlab implementation of @ejcanalesr.

**Comment by ejcanalesr** [comment]
Dear @Garyfallidis, this is a list of properties of RUMBA-SD:

1) It can be used practically with any sampling scheme, from a single shell, multiple shells, sparse q-sampling, to a cartesian DSI-grid.

2) It is among the few SD methods adding spatial regularization (Total Variation along the fibers) within the fitting. This helps to improve the angular resolution for resolving fiber crossings with small inter-fiber angles (if close to the fiber crossing area there are single fiber regions to propagate the anisotropic smoothing).

3) It is based on Rician and Non-central chi noise models, thus providing a better model for the residuals. Indeed, this algorithm is a generalization to the popular damped Richarson-Lucy SD method, but using MRI noise models and spatial regularization. 

4) In comparison to CSD, it tends to work equal or better for high b-values (b>=3000) while CSD is better for low b-values (b<2000). Some reports indicate it is more robust if a 'wrong' single fiber response function is provided. Moreover, we have a smaller number of spurious peaks/lobes because we use a discrete grid on the sphere (avoiding the sidelobes of the spherical harmonics representation near the origin). For this reason, we can use smaller thresholds to remove false positives (e.g., f>=0.1).

5) Sparse SD methods can detect fiber crossings with smaller inter-fiber angles (if the TV option is not activated) but RUMBA-SD can detect non-dominant fibers in a larger number of cases (0.1<f<0.3).

6) The global fitting is much faster than the voxelwise version, and if the TV option is used, the algorithm demands a lot of RAM (we need to have the 4D-fODF image loaded in memory for computing gradients and divergence, etc). The global fitting without the TV is very fast, however, if the TV is used the computation time can be about 1h.

7) The current implementation is based on assuming that the single response function is a diffusion tensor. However, this can be generalized in the future by considering more general models (Kurtosis, Spherical mean technique, or a set of rotated single fiber responses as determined in CSD and MTCSD, etc).

8) We found a good interaction between RUMBA-SD+TV and the probabilistic tracking in dipy. This combination won several international challenges in recent years, and we would like to promote this!

Thanks a lot for your work and for providing this fantastic resource to the community.
Please, let me know how we can help.
All the best,
Erick J Canales-Rodríguez

**Comment by ejcanalesr** [comment]
Hi again, @Garyfallidis
We have some ideas to accelerate the computation, but as this requires adding two new libraries we are not sure if we can do that. For example, the voxelwise fitting can be accelerated by using numba, and the global fitting by means of the numexpr (numexpr.evaluate) library.

Please, for more details see these links:  http://numba.pydata.org/
                                                                 https://numexpr.readthedocs.io/projects/NumExpr3/en/latest/api.html
                                                                 https://numexpr.readthedocs.io/projects/NumExpr3/en/latest/mkl.html
                                                                
Regards,
Erick

**Comment by ShreyasFadnavis** [comment]
@Garyfallidis I think this is the reference!
@kenjimarshall @gabknight @ejcanalesr : correct me if I am wrong/ there is something else we should look at?

Canales-Rodríguez EJ, Daducci A, Sotiropoulos SN, Caruyer E, Aja-Fernández S, Radua J, Yurramendi Mendizabal JM, Iturria-Medina Y, Melie-García L, Alemán-Gómez Y, Thiran JP, Sarró S, Pomarol-Clotet E, Salvador R., "Spherical Deconvolution of Multichannel Diffusion MRI Data with Non-Gaussian Noise Models and Spatial Regularization." PLoS One. 2015 Oct 15;10(10):e0138910. doi: 10.1371/journal.pone.0138910. PMID: 26470024; PMCID: PMC4607500.

**Comment by ejcanalesr** [comment]
Yes, this is the reference!
And what about "numba" and "numexpr", is it ok to add these libraries to dipy, or do you prefer the current version?
Thanks!

**Comment by ShreyasFadnavis** [comment]
> Yes, this is the reference!
> And what about "numba" and "numexpr", is it ok to add these libraries to dipy, or do you prefer the current version?
> Thanks!

I am adding @arokem and @skoudoro here... because I know that we've had multiple discussions on numba and its competitors in the past. I've had a good experience with `numba`👍🏼 but have never worked with `numexpr` though...

**Comment by ejcanalesr** [comment]
That is perfect!
numexpr is very simple, and it is fully compatible with numpy...
Thanks!
Erick

https://numexpr.readthedocs.io/projects/NumExpr3/en/latest/api.html

**Comment by matthew-brett** [comment]
Only to say that we've discussed adding Numba routines to Scipy, but
decided in the end that we would not do that, because Numba is a relatively
big install, and we didn't want to have it as a required dependency.  Just
for example, it was, at least at the time, hard to install on platforms
like Raspberry Pi.  The Numba developers agreed that was probably wise to
avoid the dependency at the time, but a) things may have changed and b)
Dipy is further up the hierarchy of dependencies, so you might have a lower
threshold to accept big dependencies.

**Comment by ShreyasFadnavis** [comment]
@ejcanalesr Can you point out to us which parts of the code are slow? Maybe we can think of Cythonizing the code and use Open MP as an alternative? I agree with @matthew-brett that Numba could be a heavy dependency which is why has not been added to DIPY yet!

**Comment by Garyfallidis** [comment]
We do have an optional dependency to TensorFlow could their be some functions there that could do the same operation as that used in numba? Also @grlee77 is developing a package using cupy. So cupy will be an optional dependency for dipy soon. 
Please also explain which parts of the code need to be optimized maybe it is trivial using Cython too.

**Comment by Garyfallidis** [comment]
So please clarify the exact parts (files/lines) that need to be optimized and justify the need using profiling tools. A simple line profiler  will do. Exciting times @ejcanalesr !

**Comment by ejcanalesr** [comment]
Great! We will highlight the files and lines that could be accelerated. 
In the global fitting function, we multiply and divide very big matrices (fODF and the diffusion data, both reshaped to 2D) many times (in a for loop). For this reason, Numpexpr could be useful to evaluate these operations in parallel (CPU). However, it would be a dream to do these operations with cupy (GPU)!

**Comment by ejcanalesr** [comment]
Hi @Garyfallidis, @ShreyasFadnavis, and @grlee77, below you can find more details:

File: dipy/reconst/rumba.py
Line 290: def rumba_deconv(data, kernel, n_iter=600, recon_type='smf', n_coils=1):
This function could be accelerated by using Numba or Cython

File: dipy/reconst/rumba.py
Function rumba_deconv_global (Lines 673-923)
Some of the lines inside this function could be accelerated with Numpexpr or Cupy.
For example, lines: 834, 839, 854, 855, 857, 858, 866, 868, 872, 873.

Next week we are going to send you the profiling results. Actually, @kenjimarshall ran some tests.
Have a nice weekend!

**Comment by arokem** [comment]
Hello! Thanks for this contribution! Looks very interesting! IMHO, I think that we can live with a numba and numexpr dependency both. I think that we could use numba in other places to speed things up, so introducing it as a dependency here, and seeing how that works could be a first foray in that direction. Certainly, numexpr is not problematic: I believe that it's a very stable and well-supported project.

Regarding the method: it seems that there are two elements to the current PR: one is the RUMBA SD method itself, with it's unique noise model and everything that goes along with that. The other element is the TV spatial regularization. I wonder how much could these two elements be decoupled? In the paper, it looks like you applied spatial regularization to both RUMBA and RL deconvolution. Could TV be applied to any deconvolution method (e.g., CSD? SHORE? ...)? Could it be applied to any reconstruction method (DTI?)?

**Comment by ejcanalesr** [comment]
Thank you, @arokem!

Next week we are going to report the computation times for both the current and accelerated versions, just to have a better idea about the improvement (and to understand if this is really important).

In theory, TV can be applied to any method. However, the current implementation only works for algorithms based on multiplicative updates (i.e., like RL: a type of gradient descent where the step sizes are derived analytically to have non-negative solutions at each iteration). So, adding TV to all the other methods is possible, but we would need to replace the current solvers (least squares, non-linear fitting, etc) with a gradient descent version where we can add the TV correction term (as the spherical harmonic coefficients can be negative, the current solver must be modified by changing the step size definition). This could be a very interesting collaboration for the near future!

We plan to participate in the DiSCO challenge organized by @gabknight: http://hardi.epfl.ch/static/events/2021_challenge/
@kenjimarshall is working hard on this, and our initial results indicate that (in the training phantoms) the combination of RUMBA-SD+TV (the current implementation in dipy) and your probabilistic tracking is producing better results than the CSD method with the same tracking. Of course, these are preliminary results, but we think we are in the right direction.

**Comment by kenjimarshall** [comment]
Dear @ShreyasFadnavis @Garyfallidis,

I hope you had great weekends!

As @ejcanalesr mentioned, I've put together some profiling results to motivate the inclusion of numexpr and/or numba.

In our `global_fit` method, much of the numerical work is done in `rumba_deconv_global` which has some very computationally expensive matrix operations when using TV. 10 iterations of RUMBA on my machine with the Stanford HARDI dataset took about 58s. Since we recommend 600 iterations, properly fitting this dataset would take about 1hr. The line profile [deconv_global_unopt.txt](https://github.com/dipy/dipy/files/6881017/deconv_global_unopt.txt) indicates that ~54% of the time is spend on large matrix operations (lines 845-850), and about 27% of the time spend on reshaping large matrices (lines 853 and 843). numexpr can be used to better manage these large matrix operations by reducing unnecessary memory allocations and splitting the large arrays into chunks. Moreover, these reshaping operations can be rewritten using numba. These changes result in a ~2x speed-up, taking about ~30s for 10 iterations. This optimized line profile is [deconv_global_opt.txt](https://github.com/dipy/dipy/files/6881019/deconv_global_opt.txt)

For reference, using the voxelwise `fit`-`odf` procedure takes ~222s for 10 iterations, about 7x slower than this optimized global fit. A line profile for `rumba_deconv` is also provided here: [deconv_voxelwise.txt](https://github.com/dipy/dipy/files/6881018/deconv_voxelwise.txt). The voxelwise fit can also be accelerated by using numba to compile `rumba_deconv`, reaching ~140s (still about 5x slower).  

In the places where numba is used, cythonizing may work just as well so we'd be happy to pursue that route if recommended. The memory management abilities of numexpr work very well with the large matrix operations in `rumba_deconv_global`, so its inclusion would complement RUMBA very well. Please let me know if you need more information!

**Comment by ShreyasFadnavis** [comment]
> Dear @ShreyasFadnavis @Garyfallidis,
> 
> I hope you had great weekends!
> 
> As @ejcanalesr mentioned, I've put together some profiling results to motivate the inclusion of numexpr and/or numba.
> 
> In our `global_fit` method, much of the numerical work is done in `rumba_deconv_global` which has some very computationally expensive matrix operations when using TV. 10 iterations of RUMBA on my machine with the Stanford HARDI dataset took about 58s. Since we recommend 600 iterations, properly fitting this dataset would take about 1hr. The line profile [deconv_global_unopt.txt](https://github.com/dipy/dipy/files/6881017/deconv_global_unopt.txt) indicates that ~54% of the time is spend on large matrix operations (lines 845-850), and about 27% of the time spend on reshaping large matrices (lines 853 and 843). numexpr can be used to better manage these large matrix operations by reducing unnecessary memory allocations and splitting the large arrays into chunks. Moreover, these reshaping operations can be rewritten using numba. These changes result in a ~2x speed-up, taking about ~30s for 10 iterations. This optimized line profile is [deconv_global_opt.txt](https://github.com/dipy/dipy/files/6881019/deconv_global_opt.txt)
> 
> For reference, using the voxelwise `fit`-`odf` procedure takes ~222s for 10 iterations, about 7x slower than this optimized global fit. A line profile for `rumba_deconv` is also provided here: [deconv_voxelwise.txt](https://github.com/dipy/dipy/files/6881018/deconv_voxelwise.txt). The voxelwise fit can also be accelerated by using numba to compile `rumba_deconv`, reaching ~140s (still about 5x slower).
> 
> In the places where numba is used, cythonizing may work just as well so we'd be happy to pursue that route if recommended. The memory management abilities of numexpr work very well with the large matrix operations in `rumba_deconv_global`, so its inclusion would complement RUMBA very well. Please let me know if you need more information!

This is not bad actually 👍🏽 Stanford HARDI is also a big subject, with a lot of volumes. If `matmul` and `reshape` is the bottleneck, definitely Cython is the way to go! Do also see this link : https://docs.scipy.org/doc/scipy/reference/special.cython_special.html ! I always find useful stuff in here, for instance Bessel function implemented here could be potentially useful :)

**Comment by ShreyasFadnavis** [comment]
> Thank you, @arokem!
> 
> Next week we are going to report the computation times for both the current and accelerated versions, just to have a better idea about the improvement (and to understand if this is really important).
> 
> In theory, TV can be applied to any method. However, the current implementation only works for algorithms based on multiplicative updates (i.e., like RL: a type of gradient descent where the step sizes are derived analytically to have non-negative solutions at each iteration). So, adding TV to all the other methods is possible, but we would need to replace the current solvers (least squares, non-linear fitting, etc) with a gradient descent version where we can add the TV correction term (as the spherical harmonic coefficients can be negative, the current solver must be modified by changing the step size definition). This could be a very interesting collaboration for the near future!
> 
> We plan to participate in the DiSCO challenge organized by @gabknight: http://hardi.epfl.ch/static/events/2021_challenge/
> @kenjimarshall is working hard on this, and our initial results indicate that (in the training phantoms) the combination of RUMBA-SD+TV (the current implementation in dipy) and your probabilistic tracking is producing better results than the CSD method with the same tracking. Of course, these are preliminary results, but we think we are in the right direction.

Sorry for being slow here... @ejcanalesr are you comparing results with Singleshell CSD or Multishell CSD? I think I remember reading in the Multishell CSD paper that 'no need for spatial regularizaion' was one of their features. Very interesting and cool to know that RUMBA+TV does so well!

**Comment by kenjimarshall** [comment]
@ShreyasFadnavis The result @ejcanalesr mentioned was using Singleshell CSD -- i.e. comparing results using one shell of the competition data with both CSD and RUMBA. RUMBA actually had better performance  in this preliminary comparison without TV as well. We haven't compared with Multishell CSD yet but we plan on doing so.

**Comment by kenjimarshall** [comment]
Some new features were just added to the PR:

1. Separate GM and CSF compartments
2. Support for single-shell, multi-shell, and recursive fiber response functions

Let me know if there are any questions!

**Comment by ShreyasFadnavis** [comment]
> Some new features were just added to the PR:
> 
> 1. Separate GM and CSF compartments
> 2. Support for single-shell, multi-shell, and recursive fiber response functions
> 
> Let me know if there are any questions!

Hi! @kenjimarshall Awesome 👍🏽 Thanks for the update. Is this PR ready for a review? Or are you still waiting to add more stuff?

**Comment by kenjimarshall** [comment]
Hi @ShreyasFadnavis - the PR is ready for review! Nothing else will be added.

**Comment by skoudoro** [comment]
Hi @kenjimarshall , Hi @ShreyasFadnavis,

We plan to talk about your PR during the DIPY online meeting tomorrow (October 7th, 2021 | 1pm EST / 7pm CET / 10am PT).

It would be great if you could attend this meeting to do a quick overview. 

More information on https://github.com/dipy/dipy/discussions/2398

Thank you!

**Comment by kenjimarshall** [comment]
Hi @skoudoro,

That's great -- I'll be there!

**Comment by arokem** [comment]
Hi Kenji. Could we meet up on Zoom on Thursday at 10am Pacific time (the same time as the DIPY meeting yesterday)? We can use this zoom link: https://washington.zoom.us/j/94691690580

**Comment by kenjimarshall** [comment]
HI @arokem I pushed these suggestions. I'm not sure why all the Linux checks were cancelled... Should I try making another commit to re-run the tests?

**Comment by arokem** [comment]
Following up here, after our discussion this morning. I think that it was suggested that it would be OK to merge this code into master, but there was some question about how to discuss the multi-shell implementation in the example. One easy way to deal with the questions around multi-shell would be to remove that example. I think that maybe @gabknight was suggesting that as a possibility. I'd alright with that. What do others think?

**Comment by ejcanalesr** [comment]
Hi Ariel/dipy team,

I agree, we could add the multi-shell example in the near future.

Thanks!

All the best,

Erick


________________________________

**Comment by arokem** [comment]
Hi @kenjimarshall : are you still available to work on this? Could you make this change? I think after removing the multi-shell example, we'd be ready to merge this.

**Comment by kenjimarshall** [comment]
Hi @arokem and everyone. Yes no problem! I'll delete it now.

**Comment by arokem** [comment]
Thanks for being so quick! I think that this is ready for the merge. I will wait until EOB Monday, to give people time to raise any objections, before I go ahead with it.

File: doc/examples/reconst_rumba.py [hunk 1]
```diff
@@ -0,0 +1,524 @@
...
+
+"""
+
+scene.rm(response_actor)
+
+"""
+**Strategy 2: estimate from local brain region**
+The `csdeconv` module contains functions for estimating this response.
+`auto_response_sst` extracts a ROI in the center of the brain and isolates
+single fiber populations from the corpus callosum using a FA mask with a
```
Comment (Garyfallidis) [inline doc/examples/reconst_rumba.py:121]: an

**Comment by Garyfallidis** [comment]
Hi @kenjimarshall and all. This PR needs rebasing.

**Review by Garyfallidis — CHANGES_REQUESTED** [review]
The PR needs a careful rebasing and a few more comments need addressing too. Overall, a great work in the making :) This is super close to be merged.

**Comment by Garyfallidis** [comment]
@arokem have you checked that the documentation of this PR renders correctly?

File: dipy/reconst/rumba.py [hunk 2]
```diff
@@ -0,0 +1,1081 @@
...
+        components are fODF while last two are GM and CSF respectively.
+
+    Notes
+    -----
+    The diffusion MRI signal measured at a given voxel is a sum of
+    contributions from each intra-voxel compartment, including parallel white
+    matter (WM) fiber populations in a given orientation as well as effects
+    from GM and CSF. The equation governing these  contributions is:
+
+    $ S_i = S_0\left(\sum_{j=1}^{M}f_j\exp(-b_i\bold{v}_i^T\bold{D}_j
```
[the remark this answers is not in the document]
Comment (kenjimarshall) [reply dipy/reconst/rumba.py]: Thanks for pointing this out -- there were some rendering issues that I believe I've now fixed.

File: dipy/reconst/tests/test_rumba.py [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+    '''
+
+    sphere = default_sphere
+
+    btable = np.loadtxt(get_fnames('dsi515btable'))
+    bvals = btable[:, 0]
+    bvecs = btable[:, 1:]
+    gtab = gradient_table(bvals, bvecs)
+
+    rumba = RumbaSDModel(gtab, n_iter=600, sphere=sphere)
```
[the remark this answers is not in the document]
Comment (kenjimarshall) [reply dipy/reconst/tests/test_rumba.py:105]: Yes I believe so, but maybe @ejcanalesr can confirm?

File: dipy/reconst/tests/test_rumba.py [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+
+    # Simulated data
+    data = single_tensor(gtab, S0=1, evals=rumba.wm_response)
+    rumba_fit = rumba.fit(data)
+    data_pred = rumba_fit.predict()
+
+    # Assert reconstructed signal value within 0.01 of original
+    assert_allclose(data, data_pred, atol=0.01, rtol=1.0)
```
[the remark this answers is not in the document]
Comment (kenjimarshall) [reply dipy/reconst/tests/test_rumba.py]: I just increased the precision on this test a bit further. It's now close to the threshold for RUMBA's predict performance.

**Comment by Garyfallidis** [comment]
Thank you for the great work @kenjimarshall and team. I am merging this now.

**Comment by Garyfallidis** [comment]
:tada:

**Pull request merged**